cclib/method/bader.py

```diff
@@ -0,0 +1,204 @@
+    ],
+    dtype=float,
+)
+
+
+class MissingInputError(Exception):
+    pass
+
+
+def __cartesian_dist(pt1, pt2):
+    """ Small utility function that calculates Euclidian distance between two points
+        pt1 and pt2 are numpy arrays representing a point in Cartesian coordinates """
+    return numpy.sqrt(numpy.einsum("ij,ij->j", pt1 - pt2, pt1 - pt2))
+
+
+class Bader(Method):
+    """Bader's QTAIM charges."""
+
+    # All of these are required for QTAIM charges.
+    required_attrs = ("homos", "mocoeffs", "nbasis", "gbasis")
+
+    def __init__(self, data, volume, progress=None, loglevel=logging.INFO, logname="Log"):
+        super(Bader, self).__init__(data, progress, loglevel, logname)
+
+        self.volume = volume
+        self.fragresults = None
+
+    def __str__(self):
+        """Return a string representation of the object."""
+        return "Bader's QTAIM charges of {}".format(self.data)
+
+    def __repr__(self):
+        """Return a representation of the object."""
+        return "Bader({})".format(self.data)
+
+    def _check_required_attributes(self):
+        super(Bader, self)._check_required_attributes()
+
+    def calculate(self, indices=None, fupdate=0.05):
+        """Calculate Bader's QTAIM charges using on-grid algorithm proposed by Henkelman group
+           in doi:10.1016/j.commatsci.2005.04.010
+           
+           Cartesian, uniformly spaced grids are assumed for this function.
+           """
+
+        # First obtain charge densities on the grid
+        if len(self.data.mocoeffs) == 1:
+            self.chgdensity = electrondensity_spin(
+                self.data, self.volume, [self.data.mocoeffs[0][: self.data.homos[0]]]
+            )
+            self.chgdensity.data *= 2
+        else:
+            self.chgdensity = electrondensity_spin(
+                self.data,
+                self.volume,
+                [
+                    self.data.mocoeffs[0][: self.data.homos[0]],
+                    self.data.mocoeffs[1][: self.data.homos[1]],
+                ],
+            )
+
+        # Assign each grid point to Bader areas
+        self.fragresults = numpy.zeros(self.chgdensity.data.shape, "d")
+        next_index = 1
+
+        self.logger.info("Partitioning space into Bader areas.")
+
+        # Exclude the outermost positions
+        for xwalker in range(1, len(self.chgdensity.data) - 1):
+            for ywalker in range(1, len(self.chgdensity.data[0]) - 1):
+                for zwalker in range(1, len(self.chgdensity.data[0][0]) - 1):
+                    (xindex, yindex, zindex) = (xwalker, ywalker, zwalker)
+
+                    if self.fragresults[xindex, yindex, zindex] != 0:
+                        # index has already been assigned for this grid point
+                        continue
+                    else:
+                        listcoord = []
+                        loopcondition = True
```

Comment:
Could you take advantage of this variable's name to convey information about the condition of the loop? (choose a better name)

Reply:
I've changed this variable to `local_max_reached`, since the loop exits when the local maximum is reached. Let me know if this can be improved further.